File: 2. Structural Design Patterns/Facade/Go/main.go
package main

import "fmt"

// Facade Pattern Example in Go
// Subsystems: CPU, Memory, HardDrive; Facade: Computer

type CPU struct{}
func (CPU) Freeze(){ fmt.Println("CPU: freeze") }
func (CPU) Jump(position int64){ fmt.Println("CPU: jump to", position) }
func (CPU) Execute(){ fmt.Println("CPU: execute") }

type Memory struct{}
func (Memory) Load(position int64, data []byte){ fmt.Printf("Memory: load %d bytes at %d\n", len(data), position) }

type HardDrive struct{}
func (HardDrive) Read(lba int64, size int){ fmt.Printf("HardDrive: read %d bytes from LBA %d\n", size, lba) }

type Computer struct{ cpu CPU; mem Memory; hdd HardDrive }
func (c Computer) Start(){ c.cpu.Freeze(); c.hdd.Read(0, 1024); c.mem.Load(0, make([]byte, 1024)); c.cpu.Jump(0); c.cpu.Execute() }
func (c Computer) Shutdown(){ fmt.Println("Computer: shutdown sequence") }

func main(){ comp := Computer{}; comp.Start(); comp.Shutdown() }

File: 2. Structural Design Patterns/Facade/Rust/main.rs
// Facade Pattern Example in Rust
// Subsystems: CPU, Memory, HardDrive; Facade: Computer

struct CPU;
impl CPU { fn freeze(&self){ println!("CPU: freeze"); } fn jump(&self, position: i64){ println!("CPU: jump to {}", position); } fn execute(&self){ println!("CPU: execute"); } }

struct Memory;
impl Memory { fn load(&self, position: i64, data_len: usize){ println!("Memory: load {} bytes at {}", data_len, position); } }

struct HardDrive;
impl HardDrive { fn read(&self, lba: i64, size: usize){ println!("HardDrive: read {} bytes from LBA {}", size, lba); } }

struct Computer { cpu: CPU, mem: Memory, hdd: HardDrive }
impl Computer {
    fn new() -> Self { Self { cpu: CPU, mem: Memory, hdd: HardDrive } }
    fn start(&self){ self.cpu.freeze(); self.hdd.read(0, 1024); self.mem.load(0, 1024); self.cpu.jump(0); self.cpu.execute(); }
    fn shutdown(&self){ println!("Computer: shutdown sequence"); }
}

fn main(){ let c = Computer::new(); c.start(); c.shutdown(); }
